### DISCUSSION TITLE
Different attribute names in ORM and Pydantic models?

### DISCUSSION THREAD

**Discussion by pacificsky** [fastapi/fastapi#7742] [Questions] [answered]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description
Hello! I am building an API using FastAPI and would like to have different attribute names across my ORM models and public user-facing resources. For example, here is my ORM User model:

```
class UserOrm(Base):
    id = Column(Integer, primary_key=True)
    name = Column(String)
```

and here is my corresponding Pydantic user model:
```
class UserResource(BaseModel):
    identifier: int
    fullname: str
```

In the public API, I'd like to accept a UserResource object where the attributes are named identifier and fullname. But internally I need some glue that tells FastAPI that `identifier` on `UserResource` maps to `id` on `UserOrm`. Is this possible?

I've explored doing this differently by defining my `UserResource` with an alias on Field like this:
```
class UserResource(BaseModel):
    id: int = Field(alias="identifier")
    name: str = Field(alias="fullname")
```

This lets me accept POST requests from clients where the attributes are named `identifier` and `fullname`, and my `UserResource` request model gets correctly populated, ie `id` and `name` have the values I'd expect. However when sending an updated `UserResource` back to the client, the automatic translation from `UserOrm`'s attribute names to the expected aliased attribute names on `UserResource` does not happen.

**Comment by fribor** [comment, marked as answer]
Hi, 

the alias parameter defines the public name for the pydantic model. What happens, is that pydantic treats your field name as a private name as soon as you are using an alias. You can change that behavior by setting `allow_population_by_field_name = True` . Your pydantic schema should look like this: 

```
class UserResource(BaseModel):
    id: int = Field(alias="identifier")
    name: str = Field(alias="fullname")

    class Config:
        allow_population_by_field_name = True
```

Here is the relevant part of the manual: [Docs](https://pydantic-docs.helpmanual.io/usage/model_config/)

**Comment by pacificsky** [comment]
Ah this is perfect - I missed `allow_population_by_field_name`. 
So now I can have 2 separate pydantic models inheriting from the same base and with the same Field aliases  - one to receive requests from the public internet (where `allow_population_by_field_name = False`) and another to send responses back (where `allow_population_by_field_name` = True) to allow for auto-mapping from ORM attribute name to pydantic model private name. 

Thanks!

**Comment by tiangolo** [comment]
Thanks for the help here @fribor ! :cake: :bowing_man: 

Thanks @pacificsky for reporting back and closing the issue :+1:

**Comment by yohhei** [comment]
Even when I set `allow_population_by_field_name = True`. The response still return dict by alias. 
Example:
```python
class RecordOut(BaseModel):
    id: str = Field(alias="_id")

    class Config:
        allow_population_by_field_name = True


@app.get("/record", response_model=RecordOut)
def read_record():
    record = {"_id": "123"}
    assert "id" in RecordOut(**record).dict(by_alias=False)
    return RecordOut(**record).dict(by_alias=False)
   # Response still has "_id" field!!!!
```

Am I doing something wrong?

**Comment by yohhei** [comment]
> Even when I set `allow_population_by_field_name = True`. The response still return dict by alias.
> Example:
> 
> ```python
> class RecordOut(BaseModel):
>     id: str = Field(alias="_id")
> 
>     class Config:
>         allow_population_by_field_name = True
> 
> 
> @app.get("/record", response_model=RecordOut)
> def read_record():
>     record = {"_id": "123"}
>     assert "id" in RecordOut(**record).dict(by_alias=False)
>     return RecordOut(**record).dict(by_alias=False)
>    # Response still has "_id" field!!!!
> ```
> 
> Am I doing something wrong?

OK, I find the solution: just add a param `response_model_by_alias=False` to the decorator.

**Comment by followben** [comment]
Adding `response_model_by_alias=False` to the decorator works, but the OpenAPI docs don't respect it (they're always generated with by_alias=True). Am I missing something @tiangolo, i.e. is this by design (and I need to set something else?); or should I create a new issue/ PR?

**Comment by ryanisn** [comment]
use `populate_by_name` in pydantic 2.0
